[SIL Diagnostics] Create a mandatory pass to check correct usage of yields in generalized accessors: _read and _modify

ravikandhadai opened the pull request:
This pass is based on the existing SIL verifier checks but diagnoses only those errors that can be introduced by programmers.

<rdar://43578476>

ravikandhadai commented:
@swift-ci Please smoke test

gottesmm commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,204 @@
+    // correctness, but it could change the errors diagnosed when there are
+    // multiple errors. Breadth-first search could diagnose errors along
+    // shorter paths.
+    llvm::DenseMap<SILBasicBlock*, BBState> visitedBBs;
```
As an aside. Please run this through git-clang-format. The patch is inconsistent about the location of the '*'.

gottesmm reviewed (commented):
Just a quick drive by.

ravikandhadai committed:
[SIL Diagnostics] Create a mandatory pass to check correct usage of
yields in generalized accessors: _read and _modify, which are
yield-once corountines. This pass is based on the existing SIL verifier
checks but diagnoses only those errors that can be introduced by programmers
when using yields.

<rdar://43578476>

ravikandhadai commented:
@swift-ci Please clean smoke test

rjmccall commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -342,6 +342,21 @@ WARNING(warning_int_to_fp_inexact, none,
+NOTE(previous_yield, none, "previous 'yield' was here", ())
+
+ERROR(possible_return_before_yield, none, "cannot guarantee that the accessor "
+  "'yield's a value before returning", ())
```
This probably reads better if you just don't quote `yield`.  It's a verb, like `return`, and doesn't need to be treated as super-technical jargon.

rjmccall commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,204 @@
+  static InFlightDiagnostic
+  diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag, U &&...args) {
+    return Context.Diags.diagnose(loc, diag, std::forward<U>(args)...);
+  }
```
You should consider pushing this up to `SILTransform`, but it's not really harmful to declare it here if that's common practice for other diagnostic passes.

rjmccall commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,204 @@
+          // When the successor follows a 'yield', update the successor state.
+          if (isa<YieldInst>(term)) {
+            insertResult.first->second.yieldState = YieldState::AfterYield;
+            insertResult.first->second.yieldInst = term;
```
You should consider encapsulating `BBState`: add functions to do the various state transitions (these can assert that (1) the current state is appropriate for the transition and (2) they've been provided the expected information (e.g. a yield instruction)) and functions to get the state-specific information (these can assert that the state is as expected).  This sort of approach encourages you to write better, stronger assertions and can catch problems a lot faster if the code needs to evolve.

rjmccall reviewed (commented):
Generally looks good; just a few style quibbles.

rjmccall commented:
It's possible that the bots run `sil-opt` tests with `-enable-sil-ownership` enabled but your local configuration doesn't.  I don't know why that would be true.

ravikandhadai commented:
@rjmccall Thanks for the inputs. I think I know the problem. I guess `@trivial` is removed by @gottesmm recently and I wasn't on recent master.

ravikandhadai commented:
@swift-ci Please smoke test

ravikandhadai commented:
@swift-ci Please smoke test OS X platform

ravikandhadai commented:
@rjmccall I have encapsulated the state transitions of BBState in this commit based on your suggestions. Let me know if this is what you had in mind.

rjmccall reviewed (approved):
Looks great, thanks!

devincoughlin reviewed (commented):
This looks good to me. There are some suggestions for diagnostic improvements inline and a suggestion to use an RPO traversal.

ravikandhadai commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,228 @@
+  public:
+    void updateToAfterYield(SILInstruction *yldInst) {
+      assert(yieldState == BeforeYield);
+      assert(yldInst != nullptr);
```
`yldInst` is a  parameter, and `yieldState` is the field of the struct: BBState. There is also a field called`yieldInst` in the BBState struct.

ravikandhadai commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -358,6 +358,21 @@ WARNING(warning_int_to_fp_inexact, none,
 
+
+// Yield usage errors
+ERROR(return_before_yield, none, "accessor doesn't yield before returning",())
```
Will do

ravikandhadai commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -358,6 +358,21 @@ WARNING(warning_int_to_fp_inexact, none,
+// Yield usage errors
+ERROR(return_before_yield, none, "accessor doesn't yield before returning",())
+
+ERROR(multiple_yields, none, "encountered multiple yields along "
```
Sure

ravikandhadai commented on test/SILOptimizer/generalized_accessors.swift:
```diff
@@ -0,0 +1,277 @@
+  var computed: Int {
+    mutating _read {
+      if stored > 0 {
+        return
```
Yes. In the following example, the diagnostic will go on the return statement

```_read {
      var loc = 1
       if flag {
         loc = 2
       }
       return // expected-error {{accessor doesn't yield before returning}}
    }
```
There can be only one return instruction in SIL. If the function can exit at multiple places the error will appear on the closing brace of the function definition. This is because at SIL level we only see one return instruction whose `Loc` is the closing brace of the function definition.

ravikandhadai commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -358,6 +358,21 @@ WARNING(warning_int_to_fp_inexact, none,
+
+NOTE(previous_yield, none, "previous yield was here", ())
+
+ERROR(possible_return_before_yield, none, "cannot guarantee that the accessor "
```
Okay.

ravikandhadai commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -358,6 +358,21 @@ WARNING(warning_int_to_fp_inexact, none,
+  "yields a value before returning", ())
+
+NOTE(conflicting_join, none,
+    "some paths reaching here have a yield and some don't", ())
```
Okay. Let me see how much we can recover here.

ravikandhadai commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,228 @@
+    llvm::Optional<SILInstruction *> returnInst = None;
+    ASTContext &astCtx = fun.getModule().getASTContext();
+
+    while (!worklist.empty()) {
```
There are a couple of issues here: (a) the analysis does visit a node more than one once, if there is a conflict i.e, in the case of error. It pushes an already visited successor on to the worklist, if the successor is seen in a state that is different from the state it was seen with before.  In the worst case, it processes each node twice, but that can happen only when there is an error. (b) The use of breadth-first search prioritizes shorter error paths over longer ones when there are multiple error paths. (However, an exception to this rule is that the analysis prioritizes "multiple-yield" error over a "return-without-yield" error by deferring the reporting of the latter error until the iteration completes.) I wonder whether RPO will preserve this behavior of shortest paths? E.g.  if we have a tree of the form:
       A
  B       C
            D

Then post-order is BDCA and I suppose reverse post-order is ACDB. So we see `B` at the end, whereas BFS will see it as the first successor of A?

rjmccall replied:
Is the goal of visiting conflict blocks multiple times just to improve diagnostics in some way?  It's not clear to me what it gets you at the moment.

ravikandhadai replied:
A block seen in the conflict case won't be visited twice, but a block seen in a non-conflict case could be processed again if it is seen in the conflict case. This is just to determine if we have a multiple yield error or a missing yield error.  E.g.

```
  if  (flag) {
     yield value
  }   <-- we see conflict here. And the error is that we don't yield on all paths.
  return   
```
OTOH, 
```
  if  (flag) {
     yield value
  }   <-- we see conflict here as before, but the error is that there are multiple yields along a path
  yield value
  return   
```
It is just done to distinguish between the above two scenarios. Multiple yield errors seem easier to report by pointing  to the two yields.

rjmccall replied:
That seems fair.

rjmccall commented:
Is this ready for review?  There's a lot of commented-out code.

ravikandhadai commented:
No not yet. I am working on recovering good diagnostics information for if-then-elses where one branch has a yield but other doesn't. I will ping you once it is ready for review.

ravikandhadai commented:
@eeckstein @atrick  This PR adds a mandatory pass for checking that generalized accessors (_read and _modify) yield exactly once. It would be great if you can provide your feedback on this.

@rjmccall @devincoughlin  I have planned to bring in the improvements to the diagnostics in two patches.  In this commit, I have rewritten the original yield-once analysis to in a data-flow analysis style by making the transfer functions and merge operations explicit. This refactoring makes adding improvements to the diagnostics easier. I have also added more tests in this commit.  Let me know you feedback on this commit.

Once this commit is merged, in a subsequent PR I will bring in the improvements for better diagnostics.  The improvements will mainly target  "conflict" cases where one branch yields but the other doesn't.

ravikandhadai commented:
@swift-ci Please smoke test

rjmccall commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,314 @@
+    /// (BeforeYield) or after seeing a yield (AfterYield), or in both states
+    /// (Conflict). This enum is a semi-lattice where Conflict is the top and
+    /// the merge of BeforeYield and AfterYield states is Conflict.
+    enum { BeforeYield, AfterYield, Conflict } yieldState = BeforeYield;
```
Please give `BBState` a constructor with two arguments, remove the defaulting, and clean up the factories.  It's fine to make it a private constructor, and you may need to add a `getInitialState` factory.

ravikandhadai replied:
Sure. I like that but the only thing that makes default initialization necessary is the DenseMap `bbToStateMap` used in the worklist algorithm. It has the type: `llvm::DenseMap<SILBasicBlock *, BBState>`. I can make the `BBState` optional and remove any default initialization/no-argument constructors. WDYT?

rjmccall replied:
I don't think `DenseMap` unconditionally requires its value type to be default-initializable; it's only required if you use methods that need default-initialization, like where you use `try_emplace`.  I could be mistaken, of course.

ravikandhadai replied:
Yes your are right, without those operations, DenseMap doesn't need default constructors. I have pushed a commit to fix these.

rjmccall replied:
Looks good, thanks!

rjmccall commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,314 @@
+    llvm::DenseMap<SILBasicBlock *, BBState> bbToStateMap;
+    SmallVector<SILBasicBlock *, 16> worklist;
+
+    auto *entryBB = &(*fun.begin());
```
`fun.getEntryBlock()`

rjmccall commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,314 @@
+    SmallVector<SILBasicBlock *, 16> worklist;
+
+    auto *entryBB = &(*fun.begin());
+    bbToStateMap.try_emplace(entryBB);
```
Having a `BBState::getInitialState()` will make this much clearer.

ravikandhadai commented:
Fixing merge conflicts.

ravikandhadai commented:
@swift-ci please smoke test

gottesmm commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,313 @@
+  private:
+    // The following state is maintained for emitting diagnostics.
+
+    // For AfterYield and Conflict states, this field records the yield
```
This should be a doxygen comment. (doxygen comment == '///'). We really need to get some sort of doxygen producing job up. It will make it easier to make sure this works well.

gottesmm commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,313 @@
+    // instruction that was seen while propagating the state.
+    SILInstruction *yieldInst;
+
+    BBState(YieldState yState, SILInstruction *yieldI):
```
Can you make sure to run this through git-clang-format before committing? (I am guessing this needs to be run through b/c I don't think git-clang-format puts the ':' there.

gottesmm commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,313 @@
+  /// A structure that records an error found during the analysis along with
+  /// some context information that will be used by diagnostics.
+  struct YieldError {
+    // The kind of error.
```
This should be a doxygen comment.

gottesmm commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,313 @@
+  /// an error that is detected.
+  /// \return the state at the exit of the basic block if it can be computed
+  /// and None otherwise.
+  llvm::Optional<BBState>
```
We have imported Optional into the swift namespace, so I don't think you need the llvm:: any more for it. On another note, can you also make this helper function have static on it?

gottesmm commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,313 @@
+
+      auto nextState = resultState.getValue();
+
+      for (auto &succ : bb->getSuccessors()) {
```
Do you ever actually need access to the successor? If not you can use:

```
for (auto *succBB : bb->getSuccessorBlocks()) {
```

ravikandhadai replied:
Sure. Will do.

gottesmm reviewed (commented):
A quick drive by. Also thanks for adding a .sil and a .swift test! = ).

ravikandhadai commented:
@swift-ci Please smoke test

ravikandhadai commented:
@swift-ci Please smoke test

ravikandhadai commented:
@swift-ci Please smoke test

ravikandhadai commented:
@swift-ci Please smoke test linux platform

shahmishal commented:
Github is having issues, which is causing CI to fail. 

https://www.githubstatus.com

<img width="998" alt="screen shot 2019-02-19 at 10 18 13 am" src="https://user-images.githubusercontent.com/2727770/53037713-be295d80-342f-11e9-847b-84cac0471459.png">

devincoughlin commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,310 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
```
Could you put a brief description here of the high-level property that the pass is checking for here and a brief overview of high-level approach it takes. Not too much detail, but enough that someone new to the project could read it and know what the purpose of this pass is.

devincoughlin commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,310 @@
+  transferStateThroughBasicBlock(SILBasicBlock *bb, BBState inState,
+                                 Optional<YieldError> &error) {
+    auto *term = bb->getTerminator();
+
```
Should this initialize 'error' to None? That way if a caller reuses the passed in error it will get correctly set to None when there is no error.

devincoughlin commented on lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp:
```diff
@@ -0,0 +1,310 @@
+      return oldState;
+    }
+
+    // Here, one state is AfterYield and the other one is BeforeYield.
```
Can you add an assertion for this? That way if the lattice changes but the merge function doesn't we'll get an assertion failure.

devincoughlin reviewed (approved):
LGTM. Some minor comments inline.

shahmishal commented:
@swift-ci Please smoke test macOS platform

ravikandhadai commented:
@devincoughlin  Thanks for the suggestions. I pushed a commit that fixes them.

eeckstein reviewed (approved):
lgtm

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge
